packages/scm/src/browser/scm-input.ts

```diff
@@ -32,6 +33,7 @@ export interface ScmInputValidator {
 export interface ScmInputOptions {
     placeholder?: string
     validator?: ScmInputValidator
+    providerContainer?: interfaces.Container
```

Comment:
exposing container is not good, why injection cannot be used?

Reply:
I do not know of a better way of supporting the requirements stated in the third paragraph in the initial PR comments.  To give flexibility going forward, I want to support the following:

- an extension package can add support for the 'history' api when the package contributing the scm did not provide a 'history' api.
- an extension package can define a new api and be a consumer of that api.  It can implement the api for specific SCMs, or downstream packages can implement the API for a SCM.

There must be a way of getting the container for the SCM type, otherwise packages can't add implementations to it.  Neither the container nor the history implementation can be injected into the ScmHistoryWidget because it will change as the user switches between repositories of different SCM types.  Let me know if you can think of a better way of doing this.
 
  three

Reply:
There can be something like:
```ts
interface ScmRepositoryFactory {
    readonly id: string;
    get(option: ScmRepositoryOptions): ScmRepository
}
```

get method will create a child container, inject ScmRepositoryOptions, different extensions can contribute different factories

not sure about id, it probably cannot be scm provider id, but should be more something like `git` or `plugin`, `plugin` will handle all scm providers contributed by VS Code extensions

Reply:
But it's not about getting the ScmRepository, it's about getting the specific API implementations.  Eg. an implementation of ScmHistorySupport, or perhaps the implementation to support the 'Amend' button.  So the 'get' function would have to be:
```ts
    get<T>(option: ScmRepositoryOptions, serviceIdentifier: interfaces.ServiceIdentifier<T>): T | undefined;
```
and then the implementation would have to use the id to find the container that has the bindings appropriate for the SCM type.  The view, when notified of a new selected ScmRepository, would have to get the id from the ScmRepoistory, then that would be looked up in a map to find the container with the bindings for that SCM type (or as ServiceIdentifier can be string as well as symbol, use the id to bind the container).  That would mean looking up the scm type container by id every time.  Then it would have to create a grandchild container (a child of the scm type container to contain the ScmRepositoryOptions), but it would have to recreate the grandchild container for each API.  This seems overly complex.

I could change it so we pass a callback to get API implementations, instead of passing the container.

Reply:
@akosyakov  I've pushed https://github.com/eclipse-theia/theia/pull/6381/commits/5bedb57f988aa789fc9339f9c23851aba8c1004c so that now the factory is injected.  Let me know if this is ok.